rust/kernel/sync/mod.rs

```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Synchronisation primitives.
+//!
+//! This module contains the kernel APIs related to synchronisation that have been ported or
+//! wrapped for usage by Rust code in the kernel and is shared by all of them.
+
+mod guard;
+mod mutex;
+
+pub use guard::{Guard, Lock};
+pub use mutex::Mutex;
+
+/// Safely initialises an object that has an `init` function that takes a name and a lock class as
+/// arguments, for example, [`Mutex`].
+#[macro_export]
+macro_rules! init_with_lockdep {
+    ($obj:expr, $name:literal) => {{
+        static mut CLASS: core::mem::MaybeUninit<$crate::bindings::lock_class_key> =
+            core::mem::MaybeUninit::uninit();
+        // SAFETY: `CLASS` is never used by Rust code directly; the kernel may change it though.
+        unsafe { $obj.init($crate::cstr!($name), CLASS.as_mut_ptr()) };
```

Comment:
Can we do something to assert the type of `$obj` -- maybe add an `_: &TYPE: &$obj;`  or something? Otherwise this feels a bit dangerous.

Reply:
This is meant to work on multiple different and unrelated types, for example, `Mutex`, `SpinLock` and `Condvar`.

How is this dangerous? (It's a genuine question, I don't quite get the dangerous aspect.)

I suppose I could define a trait that needs to be implemented by structs with such a pattern, but I'd like to understand what we gain from it. At the moment, if this macro is called on an object that doesn't have `init` or where `init` has a different type, it will just fail to compile.

Do you know of better approaches?

Reply:
So, the danger would be something like: you have a method called init on an unrelated type, which has the same signature, but different invariants. This would let you call that without `unsafe` appearing in your source code. It's a bit theoretical, I'll grant you.

I think poor error messages is probably a bigger issue in practice :-)

Reply:
I thought the macro was being used to generate the `static`. If a trait would do, it'd be better than a macro.

Reply:
Anyhow, if the macro is kept like this, we should note the macro is "unsafe" and have the usual comment (I am doing something similar for the printing stuff I am uploading).

Reply:
I see the danger now. Indeed we need to protect users from this.

@ojeda -- we need the macro to generate the static class, the trait would be in addition to the macro. It would be used to ensure that this only works when called on objects of the right type.